src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt

```diff
@@ -228,6 +231,148 @@ class UserGroupServiceTest {
     }
   }
 
+  @Nested
+  inner class AddGroupByUserId {
+
+    @Test
+    fun blankGroupCode() {
+      whenever(userRepository.findById(org.mockito.kotlin.any())).thenReturn(Optional.of(createSampleUser(username = "user")))
+      assertThatThrownBy {
+        service.addGroupByUserId(
+          UUID.fromString("00000000-aaaa-0000-aaaa-0a0a0a0a0a0a"),
+          "        "
+        )
+      }.isInstanceOf(UserGroupException::class.java)
+        .hasMessage("Add group failed for field group with reason: notfound")
+    }
+
+    @Test
+    fun groupAlreadyOnUser() {
+      val group = Group("GROUP_LICENCE_VARY", "desc")
+      val user = createSampleUser(username = "user", groups = setOf(group))
+      whenever(userRepository.findById(org.mockito.kotlin.any())).thenReturn(Optional.of(user))
+      whenever(groupRepository.findByGroupCode(anyString())).thenReturn(group)
+      assertThatThrownBy {
+        service.addGroupByUserId(
+          UUID.fromString("00000000-aaaa-0000-aaaa-0a0a0a0a0a0a"),
+          "LICENCE_VARY"
+        )
+      }.isInstanceOf(UserGroupException::class.java)
+        .hasMessage("Add group failed for field group with reason: exists")
+    }
+
+    @Test
+    fun success() {
+      whenever(authenticationFacade.currentUsername).thenReturn("MANAGER")
+      whenever(authenticationFacade.authentication).thenReturn(authentication)
+      whenever(authentication.authorities).thenReturn(SUPER_USER)
+
+      val user = createSampleUser(username = "user", groups = setOf(Group("GROUP_JOE", "desc")))
+      whenever(userRepository.findById(org.mockito.kotlin.any())).thenReturn(Optional.of(user))
+      val group = Group("GROUP_LICENCE_VARY", "desc")
+
+      val roleLicence = Authority("ROLE_LICENCE_VARY", "Role Licence Vary")
+      val roleJoe = Authority("JOE", "Role Joe")
+      group.assignableRoles.addAll(
+        setOf(
+          GroupAssignableRole(roleLicence, group, true),
+          GroupAssignableRole(roleJoe, group, false)
+        )
+      )
+      whenever(groupRepository.findByGroupCode(anyString())).thenReturn(group)
+      service.addGroupByUserId(UUID.fromString("00000000-aaaa-0000-aaaa-0a0a0a0a0a0a"), "GROUP_LICENCE_VARY")
+      assertThat(user.groups).extracting<String> { it.groupCode }.containsOnly("GROUP_JOE", "GROUP_LICENCE_VARY")
+      assertThat(user.authorities).extracting<String> { it.roleCode }.containsOnly("LICENCE_VARY")
```

Comment:
Should test the telemetry event here - and same for successAsGroupManager

Reply:
Added